### PR TITLE
adds encryption state controller

### PR DESCRIPTION
**Pull request by p0lyn0mial**
responsible for creating a single secret in openshift-config-managed with the name destName.
This single secret contains the complete EncryptionConfiguration that is consumed by the API
server that is performing the encryption.

### CODE REVIEW

File: pkg/operator/encryption/state_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,757 @@
+package encryption
+
+import (
+	"errors"
+	"fmt"
+	operatorv1 "github.com/openshift/api/operator/v1"
```
Comment (p0lyn0mial) [inline pkg/operator/encryption/state_controller_test.go]: group imports

File: pkg/operator/encryption/controllers.go [hunk 2]
```diff
@@ -18,11 +18,15 @@ import (
...
 	apiserverCodecs = serializer.NewCodecFactory(apiserverScheme)
+
+	encoder runtime.Encoder
 )
 
 func init() {
 	utilruntime.Must(apiserverconfigv1.AddToScheme(apiserverScheme))
 	utilruntime.Must(apiserverconfig.AddToScheme(apiserverScheme))
+
+	encoder = apiserverCodecs.LegacyCodec(apiserverconfigv1.SchemeGroupVersion)
```
Comment (sttts) [inline pkg/operator/encryption/controllers.go]: these are not shared traditionally, especially without more specific names. Just construct them locally.

File: pkg/operator/encryption/helpers_test.go [hunk 3]
```diff
@@ -24,6 +24,7 @@ const (
 	encryptionConfSecretForTest              = "encryption-config"
 	encryptionSecretMigratedTimestampForTest = "encryption.apiserver.operator.openshift.io/migrated-timestamp"
 	encryptionSecretMigratedResourcesForTest = "encryption.apiserver.operator.openshift.io/migrated-resources"
+	encryptionSecretReadTimestampForTest     = "encryption.operator.openshift.io/read-timestamp"
```
Comment (sttts) [inline pkg/operator/encryption/helpers_test.go]: we have a read timestamp?

File: pkg/operator/encryption/state_controller.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+// contains the complete EncryptionConfiguration that is consumed by the API
+// server that is performing the encryption.  Thus this secret represents
+// the current state of all resources in encryptedGRs.  Every encryption key
+// that matches encryptionSecretSelector is included in this final secret.
+// This secret is synced into targetNamespace at a static location.  This
+// indirection allows the cluster to recover from the deletion of targetNamespace.
+// See getResourceConfigs for details on how the raw state of all keys
+// is converted into a single encryption config.  The logic for determining
+// the current write key is of special interest.
+type encryptionStateController struct {
```
Comment (sttts) [inline pkg/operator/encryption/state_controller.go]: I skipped the `encryption` prefix everywhere.

File: pkg/operator/encryption/state_controller.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+	queue              workqueue.RateLimitingInterface
+	eventRecorder      events.Recorder
+	preRunCachesSynced []cache.InformerSynced
+
+	encryptedGRs             map[schema.GroupResource]bool
+	destName                 string
```
Comment (sttts) [inline pkg/operator/encryption/state_controller.go]: what is destName?

Comment (p0lyn0mial) [reply pkg/operator/encryption/state_controller.go]: the name of the secret that contains `EncryptionConfiguration` in `openshift-config-managed` ns

File: pkg/operator/encryption/helpers_test.go (hunk 3 above)
Comment (p0lyn0mial) [reply pkg/operator/encryption/helpers_test.go]: nope, we don't have it, thx.

File: pkg/operator/encryption/state_controller.go (hunk 5 above)
Comment (sttts) [reply pkg/operator/encryption/state_controller.go]: follow-up: ecnryptionSecretName

**Comment by sttts** [comment]
Merged as part of the prune controller PR.

**Pull request closed without merging**